[18.0][FIX] l10n_es_aeat_sii_oca: Ensure processed invoices are marked as so

AaronHForgeFlow opened the pull request:
Cuando desde la vista lista de facturas seleccionas muchas facturas y le das a enviar al SII, si alguna de ellas da un error de Odoo el resto se siguen enviando bien pero no quedan registradas en Odoo. Por tanto no sabemos cuáles se han enviado y cuáles no. Si le damos otra vez puede que AEAT nos devuelva error de factura duplicada.

Pongo el PR en borrador porque quizás la solución es un poco overkill :smile: . Abierto a sugerencias.

pedrobaeza commented on l10n_es_aeat_sii_oca/models/account_move.py:
```diff
@@ -918,11 +918,20 @@ def _send_to_sii_valid(self):
-            documents.confirm_one_document()
+        if not documents:
+            return remaining_documents
+
```
No dejes línea en blanco.

pedrobaeza commented on l10n_es_aeat_sii_oca/models/account_move.py:
```diff
@@ -918,11 +918,20 @@ def _send_to_sii_valid(self):
+            except Exception as e:
+                _logger.info(f"SII Sending failed: {e}")
+                remaining_documents |= doc
+                self.env.cr.rollback()
```
Esto no vale, porque echará atrás todo. Debes activar un savepoint para cada envío.

pedrobaeza commented on l10n_es_aeat_sii_oca/models/account_move.py:
```diff
@@ -918,11 +918,18 @@ def _send_to_sii_valid(self):
+                with self.env.cr.savepoint():
+                    doc.confirm_one_document()
+            except Exception as e:
+                _logger.info(f"SII Sending failed: {e}")
```
Esto se debería loguear el error en el propio documento, ¿no?

pedrobaeza commented on l10n_es_aeat_sii_oca/models/account_move.py:
```diff
@@ -918,11 +918,18 @@ def _send_to_sii_valid(self):
+                    doc.confirm_one_document()
+            except Exception as e:
+                _logger.info(f"SII Sending failed: {e}")
+                remaining_documents |= doc
```
Por qué repetirlo si el error puede ser por causas tuyas?

pedrobaeza commented on l10n_es_aeat_sii_oca/models/account_move.py:
```diff
@@ -918,11 +918,17 @@ def _send_to_sii_valid(self):
+                with self.env.cr.savepoint():
+                    doc.confirm_one_document()
+            except Exception as e:
+                _logger.warning(f"SII sending failed: {doc.name}: {e}")
```
Lo dicho, ¿no se debería guardar el error en el propio documento? Echa un vistazo a cómo se hace en el SII en 16:

https://github.com/OCA/l10n-spain/blob/b1d71ec1b9b8fb0ff84b6c6bef8eec26ed45418d/l10n_es_aeat_sii_oca/models/account_move.py#L664-L678

AaronHForgeFlow replied:
Cierto! Gracias!!

AaronHForgeFlow committed:
[FIX] l10n_es_aeat_sii_oca: Ensure processed invoices are updated:
When one invoice of the batch failed then we have to make sure all the others
are updated as well. Failed ones are marked as errored.

pedrobaeza reviewed (approved):
Gracias por el parche

/ocabot merge patch
